applications/StructuralMechanicsApplication/custom_elements/membrane_element.hpp

```diff
@@ -0,0 +1,433 @@
+//  Main authors:    Klaus B. Sautter
+//
+
+#if !defined(MEMBRANE_ELEMENT_3D_H_INCLUDED )
+#define  MEMBRANE_ELEMENT_3D_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/element.h"
+#include "custom_utilities/structural_mechanics_math_utilities.hpp"
+
+namespace Kratos
+{
+
+  class MembraneElement
+    : public Element
+  {
+  public:
+
+    // Counted pointer of MembraneElement
+    KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION(MembraneElement);
+
+    // Constructor using an array of nodes
+    MembraneElement(IndexType NewId, GeometryType::Pointer pGeometry);
+
+    // Constructor using an array of nodes with properties
+    MembraneElement(IndexType NewId, GeometryType::Pointer pGeometry, PropertiesType::Pointer pProperties);
+
+    // Destructor
+    ~MembraneElement() = default;
+
+
+
+    enum class VoigtType {
+      Strain,
+      Stress
+    };
+
+    enum class ConfigurationType {
+      Current,
+      Reference
+    };
+
+    enum class WrinklingType {
+      Taut,
+      Slack,
+      Wrinkle
+    };
+
+    // Name Operations
+
+    /**
+     * @brief Creates a new element
+     * @param NewId The Id of the new created element
+     * @param pGeom The pointer to the geometry of the element
+     * @param pProperties The pointer to property
+     * @return The pointer to the created element
+     */
+    Element::Pointer Create(
+        IndexType NewId,
+        GeometryType::Pointer pGeom,
+        PropertiesType::Pointer pProperties
+        ) const override;
+
+    /**
+     * @brief Creates a new element
+     * @param NewId The Id of the new created element
+     * @param ThisNodes The array containing nodes
+     * @param pProperties The pointer to property
+     * @return The pointer to the created element
+     */
+    Element::Pointer Create(
+        IndexType NewId,
+        NodesArrayType const& ThisNodes,
+        PropertiesType::Pointer pProperties
+        ) const override;
+
+    void EquationIdVector(
+      EquationIdVectorType& rResult,
+      ProcessInfo& rCurrentProcessInfo) override;
+
+    void GetDofList(
+      DofsVectorType& ElementalDofList,
+      ProcessInfo& rCurrentProcessInfo) override;
+
+    void Initialize() override;
+
+    void CalculateLeftHandSide(
+    MatrixType& rLeftHandSideMatrix,
+    ProcessInfo& rCurrentProcessInfo) override;
+
+    void CalculateRightHandSide(
+      VectorType& rRightHandSideVector,
+      ProcessInfo& rCurrentProcessInfo) override;
+
+    void CalculateLocalSystem(
+      MatrixType& rLeftHandSideMatrix,
+      VectorType& rRightHandSideVector,
+      ProcessInfo& rCurrentProcessInfo) override;
+
+
+
+     /**
+     * @brief Calculates the covariant base vectors
+     * @param rBaseVectors The base vectors to be calculated
+     * @param rShapeFunctionGradientValues Current shapefunction gradients
+     * @param Configuration Reference/Current
+     */
+    void CovariantBaseVectors(array_1d<Vector,2>& rBaseVectors,
```

Comment:
all functions except the ones in the BaseClass should be in private (or maybe protected if you derive from it)